index.ts

```diff
@@ -0,0 +1,88 @@
+import { EventEmitter } from 'events';
+
+type Handler = (...args: any[]) => void;
+type EventMap = {
+  [k: string]: Handler | Handler[] | undefined;
+};
+
+/* istanbul ignore next */
+function _ReflectApply<T, A extends any[], R> (target: (this: T, ...args: A) => R, receiver: T, args: A): R {
+  return Function.prototype.apply.call(target, receiver, args);
+}
+
+const R = typeof Reflect === 'object'
+  ? Reflect
+  /* istanbul ignore next */
+  : null;
+
+const ReflectApply = R && typeof R.apply === 'function'
+  ? R.apply
+  /* istanbul ignore next */
+  : _ReflectApply;
+
+function safeApply<T, A extends any[]> (handler: (this: T, ...args: A) => void, context: T, args: A): void {
+  try {
+    ReflectApply(handler, context, args);
+  } catch (err) {
+    // Throw error after timeout so as not to interrupt the stack
+    setTimeout(() => {
+      throw err;
+    });
+  }
+}
+
+function arrayClone<T> (arr: T[], n: number): T[] {
+  const copy = new Array(n);
+  for (let i = 0; i < n; ++i) {
+    copy[i] = arr[i];
+  }
+  return copy;
+}
```

Comment:
I don't like how `n` is an argument here. The array already contains information about its length. Let's take this opportunity to get rid of it?
```suggestion
function arrayClone<T> (arr: T[]): T[] {
  const copy = new Array(arr.length);
  for (let i = 0; i < arr.length; ++i) {
    copy[i] = arr[i];
  }
  return copy;
}
```